bin/oneoffs/migrate_storage.py

```diff
@@ -0,0 +1,216 @@
+            os.rmdir(_dirpath)
+
+
+def get_collections_files():
+    _files = []
+
+    for collection, prefix in COLLECTIONS_PREFIXES:
+        cursor = config.db.get_collection(collection).find({})
+        for document in cursor:
+            for f in get_files_by_prefix(document, prefix):
+                f_dict = {
+                    'collection_id': document.get('_id'),
+                    'collection': collection,
+                    'fileinfo': f,
+                    'prefix': prefix
+                }
+                _files.append(f_dict)
+
+    return _files
+
+
+def get_gears_files():
+    cursor = config.db.get_collection('gears').find({})
+    _files = []
+
+    for document in cursor:
+        if document.get('exchange', {}).get('git-commit', '') == 'local':
+            f_dict = {
+                'gear_id': document['_id'],
+                'gear_name': document['gear']['name'],
+                'exchange': document['exchange']
+            }
+            _files.append(f_dict)
+
+    return _files
+
+
+def migrate_collections():
+    log.info('Migrate collection files...')
+
+    _files = get_collections_files()
+
+    for i, f in enumerate(_files):
+        try:
+            file_id = f['fileinfo'].get('_id', '')
+            if file_id:
+                file_path = util.path_from_uuid(file_id)
+                if not config.fs.isfile(file_path):
+                    """Copy file from legacy to new storage"""
+
+                    log.debug('copy file between the legacy and new storage: %s' % file_path)
+
+                    dst_dir = fs.path.dirname(file_path)
+                    config.fs.makedirs(dst_dir, recreate=True)
+                    fs.move.copy_file(src_fs=config.legacy_fs, src_path=file_path, dst_fs=config.fs, dst_path=file_path)
+            else:
+                """generate uuuid, set the id field and copy the file"""
+                file_id = str(uuid.uuid4())
+                f_old_path = util.path_from_hash(f['fileinfo']['hash'])
+                f_new_path = util.path_from_uuid(file_id)
+
+                log.debug('copy file %s to %s' % (f_old_path, f_new_path))
+
+                dst_dir = fs.path.dirname(f_new_path)
+                config.fs.makedirs(dst_dir, recreate=True)
+                fs.move.copy_file(src_fs=config.legacy_fs, src_path=f_old_path, dst_fs=config.fs, dst_path=f_new_path)
+
+                update_set = {
+                    f['prefix'] + '.$.modified': datetime.datetime.utcnow(),
+                    f['prefix'] + '.$._id': file_id
+                }
+                log.debug('update file in mongo: %s' % update_set)
+                # Update the file with the newly generated UUID
+                config.db[f['collection']].find_one_and_update(
+                    {'_id': f['collection_id'], f['prefix'] + '.name': f['fileinfo']['name']},
+                    {'$set': update_set}
```

Comment:
We'll need to adjust this a bit to handle when a file is replaced while the migration is underway. An example:

While iterating over all acquisition files, any outputs from gears added to new or existing acquisitions will be stored in the new storage format and location. The migration script in it's current state handles that without issue. A "feature" of our current file placers is that they _replace_ existing files in a container if the uploaded file has the same name. If that process happens after the load of all files from the acquisition collection, the migration script will move the old file from the old path to a new (and different) path and overwrite the information in the file object. This will leave the DB in a state where it looks like a dicom -> nifti gear had been rerun, generating a new file, but the nifti referenced in the DB will be the old nifti. The new nifti will still exist, but will be an unreferenced object. 

To solve this, I think we'll need to update the `query` portion of the `find_one_and_update` to not match if the object the file document points to has changed since the initial load. Maybe matching on `hash`? We wouldn't want to use `modified` because we don't care if someone updated `info` in the meantime. We could also make adjustments to updated `created` when files are replaced. 

Note: if the `modified_count` is 0, the file was replaced or removed. We'll want to clean up the file object in the new path that will go unreferenced.

Reply:
https://github.com/scitran/core/issues/878 is relevant when discussing a reset of `created` on replace.